### ISSUE TITLE
Hosted services are started even though the application is instructed to stop.

### ISSUE THREAD

**Issue by pardahlman** [dotnet/runtime#51613]
### Description

As `Microsoft.Extensions.Hosting.Internal.Host` starts, it resolves all registered instances of  `IHostedService` and starts them in sequence. If a hosted service stops the application (using `IHostApplicationLifetime.StopApplication()`), the remaining hosted services are still started (that is `StartAsync` is called), despite the fact that the application is instructed to shut down.

This can results in a combination of the following behaviors:

* If any of the remaining hosted services checks the provided cancellation token and throws an exception, it propagates and results in an unhandled exception unless the user wraps the start call in try/catch. This feels unexpected, as the stopping should be considered graceful.
* If the remaining hosted services does not check the cancellation token, time and resources may be spent starting components that will just be stopped shortly after. This does not scale well if there are "a lot" of registered hosted services that takes "long" time to start.
 
I think it would make sense to check the `IHostApplicationLifetime.ApplicationStopping` cancellation token after each started hosted service and if it is triggered, gracefully exit the application.

One could also argue that `ThrowIfCancellationRequested` should be called on the cancellation token passed to `StartAsync`  after each started service.

Looking forward hearing your thoughts on this!

**Comment by davidfowl** [comment]
What's a the scenario for gracefully shutting down in start?

**Comment by pardahlman** [comment]
Hi @davidfowl , thanks for getting back on this issue this quickly 🙏 I suspect you're not going to be enthusiastic over the reasons behind why we want to do this, but here I go...

Basically, we are calling an external API that connects to a message broker and sets up consumers and producers for it. This call may fail for a number of reasons and if so it throws an exception. We want to log that exception and then shut down the application with a non-zero exit code. Sounds like an we should just throw an exception and let it be unhandled, right? We tried that, but something (which I now suspect is the runtime?) logs unhandled exception to the console. Everything that is written to stdout is ingested to one of our log system, so the exception was logged by our code and whatever it is that writes to the console. It would not do to just rely on the "logging" written to the console, as it does not follow our JSON schema for log entries, but more importantly, we have other logging providers that pushes data to other destinations.

We considered wrappng the call to `IHost.RunAsync` in a try/catch, but at this point we have too many services in too many repositories for such change to be feasible. Based on this I think you can guess what solution we ended up with: using `StopApplication` on `IHostApplicationLifetime`.

So that is the backstory... but, if we raise the issue to a more theoretical level, wouldn't you expect the scenario described in the issue to result in a graceful shutdown?

**Comment by davidfowl** [comment]
So if I got this right, you don't want the exception to be printed to the console, *and* you don't want to add a try/catch to your code, but you want a non-zero exit code for a graceful shutdown? 

The graceful shutdown sounds reasonable, but not the non-zero exit code.

**Comment by pardahlman** [comment]
You got it right 👌 The non-zero exit code is part of another implementation detail that has to do with compatibility with services still running on Windows. It has very little bearing for this issue. What I'm suggesting is that it would be reasonable to expect the following:

1. If an `IHostedService` stops the application (`IHostApplicationLifetime.StopApplication()`) it should result in a graceful shutdown, preferably as soon as possible (e.g. check before starting each `IHostedService`).
2. If cancellation is requested for the user provided `CancellationToken` it should throw an `OperationCanceledException` as soon as possible (e.g. check before starting each `IHostedService`)

If you think this makes sense I could submit an MR for these changes.

**Comment by davidfowl** [comment]
> If an IHostedService stops the application (IHostApplicationLifetime.StopApplication()) it should result in a graceful shutdown, preferably as soon as possible (e.g. check before starting each IHostedService).

This sounds reasonable. In this case is an exception thrown? If yes, how does it fix your scenario?

> If cancellation is requested for the user provided CancellationToken it should throw an OperationCanceledException as soon as possible (e.g. check before starting each IHostedService)

Yes this is fine.

**Comment by pardahlman** [comment]
> In this case is an exception thrown? If yes, how does it fix your scenario?

No, I would suggest that the graceful shutdown does not throw an unhandled exception, but instead aborts the starting up of hosted services, probably do some cleanup (like stopping the already started services? call `ApplicationLifetime.NotifyStopped`?) and then return.

I guess there is an argument that stopping the application (`IHostApplicationLifetime.StopApplication()`) triggers a cancellation request on the `CancellationToken` `ApplicationStopping`, which would be a reason to throw `OperationCanceledException`.  I haven't considered that to be a desired behavior, as I don't think an unhandled exception from `IHost.StartAsync` is graceful. Let me know if you're seeing this differently!

**Comment by pardahlman** [comment]
I've been giving this issue some more thought and created an PR with a suggestion of how to modify the startup behavior to respect `IHostApplicationLifetime.ApplicationStopping` and the user provided `CancellationToken` while starting up the hosted services. Looking forward hearing your thought on this!

**Comment by maryamariyan** [comment]
> It would not do to just rely on the "logging" written to the console, as it does not follow our JSON schema for log entries

not completely related to the gist of this issue, but also wanted to mention that console logs can be written in a JSON format too when you use AddJsonConsole.

**Comment by karelz** [comment]
Triage: Graceful shutdown sounds reasonable, but given only 1 person was interested, it is a very low priority. We can keep it open for now.
